packages/core/lib/testing/testrunner.js

```diff
@@ -49,231 +50,164 @@ function TestRunner(options = {}) {
+    // await this.deploy();
+
+    try {
+      let initial_snapshot = await this.snapshot();
+      this.can_snapshot = true;
+      this.initial_snapshot = initial_snapshot;
+    } catch (_) {}
+    this.first_snapshot = false;
+  } else {
+    await this.resetState();
+  }
 
-          abis.map(function(abi) {
-            if (abi.type === "event") {
-              var signature =
-                abi.name + "(" + _.map(abi.inputs, "type").join(",") + ")";
-              self.known_events[web3Utils.sha3(signature)] = {
-                signature: signature,
-                abi_entry: abi
-              };
-            }
-          });
-          callback();
-        }
-      );
-    });
-  };
+  let files = await fs.readdir(this.config.contracts_build_directory);
 
-  if (self.first_snapshot) {
-    // // Make the initial deployment (full migration).
-    // self.deploy(function(err) {
-    //   if (err) return callback(err);
+  files = files.filter(file => path.extname(file) === ".json");
 
-    self.snapshot(function(err, initial_snapshot) {
-      if (err == null) {
-        self.can_snapshot = true;
-        self.initial_snapshot = initial_snapshot;
-      }
+  let data = await Promise.all(
+    files.map(
+      async file =>
+        await fs.readFile(
+          path.join(this.config.contracts_build_directory, file),
+          "utf8"
+        )
+    )
+  );
 
-      self.first_snapshot = false;
+  let contracts = data.map(JSON.parse).map(json => contract(json));
 
-      afterStateReset();
-    });
-    //});
-  } else {
-    self.resetState(afterStateReset);
-  }
+  //NOTE: in future should get rid of contracts argument
+  this.decoder = await Decoder.forProject(this.provider, contracts);
 };
 
-TestRunner.prototype.deploy = function(callback) {
-  Migrate.run(
+TestRunner.prototype.deploy = async function() {
+  await Migrate.run(
     this.config.with({
       reset: true,
       quiet: true
-    }),
-    callback
+    })
   );
 };
 
-TestRunner.prototype.resetState = function(callback) {
-  var self = this;
+TestRunner.prototype.resetState = async function() {
   if (this.can_snapshot) {
-    this.revert(this.initial_snapshot, function(err) {
-      if (err) return callback(err);
-      self.snapshot(function(err, snapshot) {
-        if (err) return callback(err);
-        self.initial_snapshot = snapshot;
-        callback();
-      });
-    });
+    await this.revert(this.initial_snapshot);
+    this.initial_snapshot = await this.snapshot();
   } else {
-    this.deploy(callback);
+    await this.deploy();
   }
 };
 
-TestRunner.prototype.startTest = function(mocha, callback) {
-  this.interfaceAdapter
-    .getBlockNumber()
-    .then(result => {
-      const one = web3Utils.toBN(1);
-      const resultBN = web3Utils.toBN(result);
-
-      // Add one in base 10
-      this.currentTestStartBlock = resultBN.add(one);
+TestRunner.prototype.startTest = async function() {
+  let blockNumber = await this.web3.eth.getBlockNumber();
```

Comment:
Should this call use `interfaceAdapter`?

Reply:
How would that work exactly?  I haven't been keeping up with that.

Reply:
I'm surprised we still even have the `Web3Shim` in here, since it doesn't look like it's being used.
```suggestion
  let blockNumber = await this.interfaceAdapter.getBlockNumber();
```

Reply:
Well, `this.web3` is a `Web3Shim`, so it is being used in that one place.  But if we change this one call like you say, then we can get rid of it.  So that's nice.  (The `web3` object/shim used to also be used for getting the logs, but now that's handled by `decoder.events()`.)